sklearn/tree/_classes.py

```diff
@@ -97,6 +99,30 @@ class BaseDecisionTree(MultiOutputMixin, BaseEstimator, metaclass=ABCMeta):
     Use derived classes instead.
     """
 
+    _parameter_constraints = {
+        "splitter": [StrOptions({"best", "random"})],
+        "max_depth": [Interval(Integral, 1, None, closed="left"), None],
+        "min_samples_split": [
+            Interval(Integral, 2, None, closed="left"),
+            Interval(Real, 0.0, 1.0, closed="neither"),
+        ],
+        "min_samples_leaf": [
+            Interval(Integral, 1, None, closed="left"),
+            Interval(Real, 0.0, 1.0, closed="neither"),
+        ],
+        "min_weight_fraction_leaf": [Interval(Real, 0.0, 0.5, closed="both")],
+        "max_features": [
+            Interval(Integral, 1, None, closed="left"),
+            Interval(Real, 0.0, 1.0, closed="right"),
```

Comment:
In the current API, they do overlap. Integer `1` and float `1.0` means two different things.

If we want to remove the semantics, then float `1.0` would be the one to remove because `max_features=None` does the same thing. The issue with removing it float `1.0` is that is makes hyperpamarmeter search spaces for `max_features` a bit harder to define if one wants to include "all features" in the search.

Reply:
I find it highly error prone to allow different behavior depending on whether you put `1` or `1.`. But let's keep the existing behavior in this PR for simplicity.